src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -77,12 +77,12 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.MUST));
                     }
                 } else if ("must_not".equals(currentFieldName) || "mustNot".equals(currentFieldName)) {
-                    Query query = parseContext.parseInnerQuery();
+                    Query query = parseContext.parseInnerFilter();
                     if (query != null) {
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.MUST_NOT));
                     }
                 } else if ("should".equals(currentFieldName)) {
-                    Query query = parseContext.parseInnerFilter();
+                    Query query = parseContext.parseInnerQuery();
                     if (query != null) {
                         clauses.add(new BooleanClause(query, BooleanClause.Occur.SHOULD));
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
One question here. I'm trying to understand how the isFilter() flag that parseInnerFilter sets is used after it is set. Should this be moved to the "must_not" clause as well?